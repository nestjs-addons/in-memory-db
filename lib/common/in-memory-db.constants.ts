export const IN_MEMORY_DB_SERVICE = 'InMemoryDBService';
export const IN_MEMORY_DB_CONFIG = 'InMemoryDBConfig';
