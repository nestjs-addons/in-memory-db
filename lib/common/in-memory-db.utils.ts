import { IN_MEMORY_DB_SERVICE } from './in-memory-db.constants';

export function getInMemoryDBServiceToken(featureName?: string) {
  return featureName && featureName !== IN_MEMORY_DB_SERVICE
    ? `${featureName}${IN_MEMORY_DB_SERVICE}`
    : IN_MEMORY_DB_SERVICE;
}
