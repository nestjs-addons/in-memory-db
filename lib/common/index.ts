export * from './in-memory-db.decorators';
export { getInMemoryDBServiceToken } from './in-memory-db.utils';
