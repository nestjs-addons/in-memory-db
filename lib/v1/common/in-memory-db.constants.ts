export const IN_MEMORY_DB_V1_SERVICE = 'InMemoryDBV1Service';
export const IN_MEMORY_DB_V1_CONFIG = 'InMemoryDBV1Config';
