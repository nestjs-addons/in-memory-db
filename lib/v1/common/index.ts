export * from './in-memory-db.decorators';
export { getInMemoryDBV1ServiceToken } from './in-memory-db.utils';
