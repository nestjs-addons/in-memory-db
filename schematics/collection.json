{
    "$schema": "../node_modules/@angular-devkit/schematics/collection-schema.json",
    "schematics": {
      "nest-add": {
        "description": "Schematics to add In memory DB module.",
        "factory": "./nest-add/",
        "schema": "./nest-add/schema.json"
      }
    }
  }
