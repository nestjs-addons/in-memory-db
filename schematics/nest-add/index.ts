import {
  Rule,
  SchematicContext,
  SchematicsException,
  Tree,
  chain,
  branchAndMerge,
} from '@angular-devkit/schematics';
import { JsonParseMode, parseJson, Path } from '@angular-devkit/core';
import {
  addPackageJsonDependency,
  NodeDependency,
  NodeDependencyType,
} from 'schematics-utilities';
import { NgAddOptions } from './add.schema';
import { ModuleFinder } from '../utils/module.finder';
import * as ts from 'typescript';
import { addImportToModule, insertImport } from '../utils/ast-utils';
import { InsertChange } from '../utils/change';


function getWorkspace(tree: Tree): { path: string; workspace: any } {
  const possibleFiles = ['/nest-cli.json'];
  const path = possibleFiles.filter(configPath => tree.exists(configPath))[0];

  const configBuffer = tree.read(path);
  if (configBuffer === null) {
    throw new SchematicsException(`Could not find nest-cli.json`);
  }
  const content = configBuffer.toString();
  let workspace: any;
  try {
    workspace = parseJson(content, JsonParseMode.Loose);
  } catch (e) {
    throw new SchematicsException(
      `Could not parse nest-cli.json: ` + e.message,
    );
  }
  return {
    path,
    workspace,
  };
}

function addDeclarationToModule(options: NgAddOptions): Rule {
  return (tree: Tree) => {

    if (options.skipImport !== undefined && options.skipImport) {
      return tree;
    }

    const { path, workspace } = getWorkspace(tree);

    options.module = new ModuleFinder(tree).find({
      name: options.module ? options.module : 'app',
      path: ('./' + workspace.sourceRoot) as Path,
    });

    if (!tree.exists(options.module)) {
      throw new SchematicsException(`Could not find root module, please use --module flag to specify the root module path.`);
    }

    const content = tree.read(options.module).toString();
    const modulePath =
      './' + workspace.sourceRoot + '/' + options.module
        ? options.module
        : 'app';

    const source = ts.createSourceFile(
      modulePath,
      content,
      ts.ScriptTarget.Latest,
      true,
    );

    const importChanges = addImportToModule(
      source,
      modulePath,
      'InMemoryDBModule.forRoot()',
      '@nestjs-addons/in-memory-db',
    ).shift();

    const commonImports = [
      insertImport(
        source,
        modulePath,
        'InMemoryDBModule',
        '@nestjs-addons/in-memory-db',
      ),
      importChanges,
    ];
    const changes = [...commonImports];
    const recorder = tree.beginUpdate(modulePath);
    for (const change of changes) {
      if (change instanceof InsertChange) {
        recorder.insertLeft(change.pos, change.toAdd);
      }
    }
    tree.commitUpdate(recorder);

    return tree;
  };
}

export default function (options: NgAddOptions): Rule {
  return branchAndMerge(
    chain([
      addDeclarationToModule(options)
    ]),
  );
}
