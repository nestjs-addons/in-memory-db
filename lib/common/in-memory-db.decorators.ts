import { Inject } from '@nestjs/common';
import { getInMemoryDBServiceToken } from './in-memory-db.utils';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const InjectInMemoryDBService = (featureName: string) =>
  Inject(getInMemoryDBServiceToken(featureName));
