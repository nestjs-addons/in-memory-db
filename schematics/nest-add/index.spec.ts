import { normalize } from '@angular-devkit/core';
import {
    SchematicTestRunner,
    UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { HostTree } from '@angular-devkit/schematics';

const collectionPath = path.join(__dirname, '../collection.json');

const MAIN_FILE = `
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  await app.listen(3000);
}
bootstrap();
`;

const APP_MODULE_CONTENT = `
import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';

@Module({
  imports: [],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
`;


describe('nest add function', () => {
    let tree: UnitTestTree;
    let runner: SchematicTestRunner;
    beforeEach(() => {
        tree = new UnitTestTree(new HostTree());
        tree.create('/package.json', JSON.stringify({}));
        tree.create('/nest-cli.json', JSON.stringify({
            "language": "ts",
            "collection": "@nestjs/schematics",
            "sourceRoot": "src"
        }));
        tree.create('/src/main.ts', MAIN_FILE);
        tree.create('/src/app.module.ts', APP_MODULE_CONTENT);
        runner = new SchematicTestRunner('schematics', collectionPath);
    })
    it('should add package to module', () => {
        const ngAddTree = runner.runSchematic('nest-add', '', tree);
        const module = ngAddTree.readContent('/src/app.module.ts');
        expect(module).toMatchSnapshot();
    });
})