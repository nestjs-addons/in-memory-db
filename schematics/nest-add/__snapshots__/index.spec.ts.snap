// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nest add function should add package to module 1`] = `
"
import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { InMemoryDBModule } from '@nestjs-addons/in-memory-db';

@Module({
  imports: [InMemoryDBModule.forRoot()],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
"
`;
